### PR TITLE
Add ClassifyCommonConversion to SemanticModel.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @jcouv

**Comment by CyrusNajmabadi** [comment]
Note: this simply mirrors what was done with Compilation, where we added Compilation.ClassifyCommonConversion to mirror Compilation.ClassifyConversion.  This just extends this to SemanticModel as well to mirror its existing Classify Conversion helpers.

**Comment by gafter** [comment]
@dotnet/compiler-api Opinions?

**Comment by CyrusNajmabadi** [comment]
@gafter @jcouv @dotnet/compiler-api Thoughts on this?

**Comment by jcouv** [comment]
I think it makes sense (it's consistent with other "Common" APIs) and we already have identified one IDE scenario that would use it.

**Comment by jcouv** [comment]
If we do go ahead, we should add some tests on the new public API though.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 1]
```diff
@@ -2470,6 +2470,23 @@ public bool TryGetSpeculativeSemanticModel(int position, AttributeSyntax attribu
             get;
         }
 
+        /// <summary>
+        /// Determines what type of conversion, if any, would be used if a given expression was
+        /// converted to a given type.  If isExplicitInSource is true, the conversion produced is
+        /// that which would be used if the conversion were done for a cast expression.
+        /// </summary>
+        /// <param name="expression">An expression which much occur within the syntax tree
```
Comment (KathleenDollard) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: Typo "much"

File: src/Compilers/Core/Portable/Compilation/SemanticModel.cs [hunk 2]
```diff
@@ -888,5 +888,41 @@ protected internal virtual SyntaxNode GetTopmostNodeForDiagnosticAnalysis(ISymbo
...
         /// </summary>
         protected abstract SyntaxNode RootCore { get; }
+
+#pragma warning disable RS0026 // Do not add multiple public overloads with optional parameters
+        /// <summary>
+        /// Determines what type of conversion, if any, would be used if a given expression was
+        /// converted to a given type.  If isExplicitInSource is true, the conversion produced is
+        /// that which would be used if the conversion were done for a cast expression.
+        /// </summary>
+        /// <param name="expression">An expression which much occur within the syntax tree
```
Comment (KathleenDollard) [inline src/Compilers/Core/Portable/Compilation/SemanticModel.cs]: typo: much

File: src/Compilers/VisualBasic/Portable/Compilation/SemanticModel.vb [hunk 3]
```diff
@@ -2313,6 +2330,23 @@ _Default:
...
+        ''' <param name="expression">An expression to classify. This expression does not need to be
+        ''' present in the syntax tree associated with this object.</param>
+        ''' <param name="destination">The type to attempt conversion to.</param>
+        ''' <returns>Returns a Conversion object that summarizes whether the conversion was
+        ''' possible, and if so, what kind of conversion it was. If no conversion was possible, a
+        ''' Conversion object with a false "Exists " property is returned.</returns>
+        ''' <remarks>To determine the conversion between two types (instead of an expression and a
+        ''' type), use Compilation.ClassifyConversion.</remarks>
+        Public Overrides Function ClassifyCommonConversion(position As Integer, expression As SyntaxNode, destination As ITypeSymbol, Optional isExplicitInSource As Boolean = False) As CommonConversion
+            Return ClassifyConversion(position, DirectCast(expression, ExpressionSyntax), destination).ToCommonConversion()
```
Comment (KathleenDollard) [inline src/Compilers/VisualBasic/Portable/Compilation/SemanticModel.vb]: `isExplicitInSource` is not used - is there a parameter missing to ClassifyConversion? (and this doesn't match C# version)

File: src/Compilers/VisualBasic/Portable/Compilation/SemanticModel.vb [hunk 4]
```diff
@@ -2299,6 +2299,23 @@ _Default:
...
+        ''' associated with this object.</param>
+        ''' <param name="destination">The type to attempt conversion to.</param>
+        ''' <param name="isExplicitInSource">Unused.</param>
+        ''' <returns>Returns a Conversion object that summarizes whether the conversion was
+        ''' possible, and if so, what kind of conversion it was. If no conversion was possible, a
+        ''' Conversion object with a false "Exists " property is returned.</returns>
+        ''' <remarks>To determine the conversion between two types (instead of an expression and a
+        ''' type), use Compilation.ClassifyConversion.</remarks>
+        Public Overrides Function ClassifyCommonConversion(expression As SyntaxNode, destination As ITypeSymbol, Optional isExplicitInSource As Boolean = False) As CommonConversion
+            Return ClassifyConversion(DirectCast(expression, ExpressionSyntax), destination).ToCommonConversion()
```
Comment (KathleenDollard) [inline src/Compilers/VisualBasic/Portable/Compilation/SemanticModel.vb]: `isExplicitInSource` is not used - is there a parameter missing to ClassifyConversion? (and this doesn't match C# version)

**Comment by CyrusNajmabadi** [comment]
closing

**Pull request closed without merging**